### ISSUE TITLE
Suggestions for viewing long-read structural variant data

### ISSUE THREAD

**Issue by jrobinso** [igvteam/igv#632]
From: https://groups.google.com/d/msgid/igv-help/3c79a4a8-f759-4380-a1d9-ff65083e99a4%40googlegroups.com?utm_medium=email&utm_source=footer

I've been using IGV lately to look at some Oxford Nanopore sequence data to find structural variants. IGV has been helpful, but there are definitely a few features (drawing on functionality for SVs from Illumina paired-end reads) which would be extra useful: 

My wishlist would be:

1. Setting to show hard-clipped bases as well as soft-clipped. 

Long-read aligners use a mixture of secondary and supplementary alignments (ie a mixture of hard and soft-clipping) to indicate split reads. These split reads are the ones that show structural variants, and the ones we want to see. Right now, IGV only allows soft-clipped bases to be shown, so that about half of the split reads don't show up their clipped portion when "show soft-clipped bases" is checked. It would be very helpful to have a setting to also show hard-clipped bases.

2. "View secondary/supplementary alignment region in split screen"

The "view mate region in split screen" functionality is invaluable for looking at SVs in Illumina reads. It would be equally invaluable to be able to do this for long split reads. One possible issue would be that long reads often have multiple additional alignments. Maybe in these cases it'd be best to open multiple views at once?

3. Extend "link supplementary alignments" to also link secondary alignments

As mentioned in 1), the choice between marking split alignments as secondary or supplementary seems to be somewhat arbitrary on the part of the aligner. At present, only supplementary alignments are being linked/coloured, which makes finding the secondary alignments tricky (since it involved hovering over reads one by one).

4. Create a "color alignments by secondary/supplementary alignment orientation"
The "color alignments by mate orientation" feature is also really useful; it would be great to have something similar for split long reads.

5. Highlight reads with clipped region greater than N
It would be very helpful to have some kind of visual cue (colour? read shape?) to find reads with large clipped sections. These sections often indicate SVs. Sometimes they are not aligned (eg if they're only a few hundred bp and in the middle of a repetitive region), but are clipped. Being able to quickly find and examine these reads would be really helpful.

**Comment by oneillkza** [comment]
Thanks for opening this. Followed. Tag me if it would be helpful to make any clarifications!

**Comment by helgathorv** [comment]
From @oneillkza at igv-help https://groups.google.com/d/msgid/igv-help/3c79a4a8-f759-4380-a1d9-ff65083e99a4%40googlegroups.com

Now that I look at it, minimap2 has a switch (-Y) to force it to
always use soft-clipping for supplementary alignments, which would
obviate this.

It sounds like 1 and 3 on my list actually aren't needed.

2,4 and 5 would still be very useful!

**Comment by oneillkza** [comment]
I also just realised that 5 exists in the "flag clipping > N bases" option. 

After some discussion with some folks about our specific data set, one other thing we thought of:

6) For linked supplementary alignments, show which ends pair up with which.
We've been looking at a lot of inversions, and it would be very helpful to be able to more easily see which ends contain the breakpoint. Right now the flag clipping > N bases" helps with this, but doesn't work if there are breakpoints at both ends of an alignment.

**Comment by oneillkza** [comment]
Oh, one more:

7. Add a "blat soft-clipped region" option, similar to "blat insert". Often there are reads lining up at a breakpoint, but the aligner didn't quite have enough information to place the clipped portion. Likely this functionality would need separate "blat left clipped sequence" and "blat right clipped sequence" options.

**Comment by amwenger** [comment]
I like the ideas here and am happy to send PRs for a few that are within my abilities to implement.  To start, I sent #728, which adds right-click menu items to BLAT clipped sequences (idea 7 above).

A few ideas on some of the other suggestions:

> 3. Extend "link supplementary alignments" to also link secondary alignments

I like that idea.  Perhaps the option could be renamed "link alignments" and would link all alignments with the same read name.  I usually filter secondary alignments, but it makes sense to link them if they are present.

You can also use "Group by > ZMW" to accomplish something similar for PacBio reads and group all alignments of a read together.

> 4. Create a "color alignments by secondary/supplementary alignment orientation" The "color alignments by mate orientation" feature is also really useful; it would be great to have something similar for split long reads.

One way to accomplish this would be to use different colors for the ends of clipped alignments (currently always magenta) based on how the clipped portion of the read aligns:
* rest of the read does not align
* rest of the read aligns downstream in the reference on the same strand without skipping read sequence (DELETION)
* rest of the read aligns upstream in the reference on the same strand without skipping read sequence (DUPLICATION)
* rest of the read aligns nearby in the reference on the same strand and skips read sequence (INSERTION)
* rest of the read aligns nearby in the reference on the opposite strand without skipping read sequence (INVERSION)
* rest of the read aligns far away in the reference (TRANSLOCATION)

> 6. For linked supplementary alignments, show which ends pair up with which.

Different colors could be used for the end of clipped alignments to indicate matched ends (i.e. green end goes with green end, blue end with blue end, etc.).

Another idea is to sort alignments vertically by read start coordinate when "Group by ZMW" is used.  That way adjacent portions of the read would be on consecutive lines in the group.

**Comment by jrobinso** [comment]
@amwenger re linking secondary alignments, by the spec these are not split (chimeric) reads, rather they are alternative alignments for the whole read sequence.  So linking them seems odd.  The OP indicates that aligners sometimes mix these terms up and do the wrong thing,  is that still true?  Its easy enough to add the option to cover misbehaving alignments, but they are fundamentally a different thing.

WRT using sort and group by ZMW together, sorting sorts rows based on the reads that intersect the center line, it does not reassign reads to different rows.  So what you state is true only for those reads in the center.  Just clarifying, not sure if that's what you meant.  

I suggest we make a new issue for the color-by-orientation idea as it is atomic and can be accomplished, unlike the current issue.   I will do this and copy your text, but if you want to track it its probably better if you do it.

**Comment by oneillkza** [comment]
@jprobinso, yep I'm pretty sure you're right, and I was mistaken about needing features for secondary alignments. Mostly I had been confused about hard/soft clipping, which as mentioned earlier I can resolve at the aligner level.

It might still make sense to have some means of linking secondary alignments, but in general I'd expect that these would be cases of ERVs and similar repeats, which would rarely be present in the same viewer.

Mainly the features that would still be useful for long-reads SVs would be "view supplementary alignment region in split screen" and some way of colouring reads with supplementary alignments.

PS: the "third gen" options in 2.7 have been very useful. Thank you!

**Comment by oneillkza** [comment]
See just above for the split off feature request! #731

**Comment by jrobinso** [comment]
@oneillkza OK, got it.  I'll pick this up in the next round of IGV development, post thanksgiving.
